framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -195,9 +198,48 @@ private void viewpointAdaptExecutable(
             receiver = FlowExpressions.internalReprOf(factory, receiverTree);
         }
 
-        List<FlowExpressions.Receiver> argReceivers = new ArrayList<>(args.size());
-        for (ExpressionTree argTree : args) {
-            argReceivers.add(FlowExpressions.internalReprOf(factory, argTree));
+        List<Receiver> argReceivers = new ArrayList<>();
+        boolean isVarargs = false;
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+            ExecutableElement methodCalled = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+            if (methodCalled != null && methodCalled.isVarArgs()) {
+                isVarargs = true;
+                for (int i = 0; i < methodCalled.getParameters().size() - 1; i++) {
+                    argReceivers.add(FlowExpressions.internalReprOf(factory, args.get(i)));
+                }
+                List<Receiver> varargArgs = new ArrayList<>();
+                for (int i = methodCalled.getParameters().size() - 1; i < args.size(); i++) {
+                    varargArgs.add(FlowExpressions.internalReprOf(factory, args.get(i)));
+                }
+                Element varargsElement =
+                        methodCalled.getParameters().get(methodCalled.getParameters().size() - 1);
+                TypeMirror tm = ElementUtils.getType(varargsElement);
+                boolean passedArray = false;
+                if (methodCalled.getParameters().size() == args.size()) {
+                    AnnotatedTypeMirror lastArg =
+                            factory.getAnnotatedType(args.get(args.size() - 1));
+                    List<AnnotatedTypeMirror> argTypes = typeFromUse.getParameterTypes();
+                    AnnotatedArrayType varargsArg =
+                            (AnnotatedArrayType) argTypes.get(argTypes.size() - 1);
+                    if (lastArg.getKind() == TypeKind.ARRAY
+                            && AnnotatedTypes.getArrayDepth(varargsArg)
+                                    == AnnotatedTypes.getArrayDepth((AnnotatedArrayType) lastArg)) {
+                        argReceivers.add(varargArgs.get(0));
+                        passedArray = true;
```

Comment:
It seems that the case of `passedArray == true` has the same result as the case `isVarargs == false`.
I'd recommend moving these two ifs above to the point where `isVarargs = true` is, and perhaps moving all these checks to a helper method `isVarArgsInvocation(methodCalled, typeFromUse, args)` or similar. That would separate the two things this code is doing - (1) checking if it is a varargs invocation and (2) filling the `argReceivers` list - and make it easier to understand.